### PR TITLE
Snapshot restore clears the session gate; redacted alias creds and unmatched mounts are reported

### PR DESCRIPTION
**Pull request by bytecii**
Three snapshot fixes from the 2026-09-05 issue batch, plus the doc corrections #1024 asked for, in both languages.

## Restored env clears the session gate (#1017)

`_restore_sessions` / `restoreSessions` seeded `session.vars` directly and `restore_seed` installed the env template the same way, so a `pre_session` rule refusing a name (`AWS_*`, a credential) was advisory for the one env input a deployment did not author. A new `gate_restored_vars` / `gateRestoredVars` (beside `set_var`, whose value rendering it shares through `gate_rendering`) fires the gate for every restored variable and the template, as a `set` of the rendered value, before any of them lands. A refusal aborts the load with the `PolicyDenied` a live `export` reports rather than dropping one variable: a partial restore is a workspace whose state matches no snapshot. The shell's own bookkeeping (`PWD`, `OLDPWD`, `BASH_REMATCH`, the names `seed_var` writes ungated live) is exempt on restore as it is live. This reaches every door that applies a state dict: `from_state`, the server's version restore/checkout, and the OpenAI sandbox.

## A redacted alias credential is caught, and an unmatched mount is reported (#1019)

- `has_redacted_secret` scanned only the secret fields of the config class the saved mount resolved to. An alias resource (MinIO) saves its own config under its parent's `type` (`s3`), so the class named the wrong fields, the load proceeded, and the mount rebuilt with the literal `<REDACTED>` as its key (a 403 on first read). It now scans every value, as TypeScript's `hasRedactedSecret` always did; `requires_resource_override` for a MinIO snapshot is `True` and `build_mount_args` names `/s3/` in its refusal.
- A snapshot mount with no mount at that exact prefix is still never resolved to an ancestor, but the skip is now a warning naming the prefix (Python `logger.warning`, TS `console.warn`) instead of silence.

## Docs that promised more than the code held (#1024)

- CLAUDE.md's session-gate bullet now reads "every session write a shell line names", lists the two exempt host-side doors (shell bookkeeping through `seed_var`, the embedder's `SessionManager.env`), notes that a `for` variable is not written back at all, and says the restore is not exempt.
- The history bullet now says what the two views are: the builtin is one session's list after its `CLEAR`/`DELETE` events, the file is the shared histfile of every session (as `~/.bash_history` is; `history -c` clears the list, not the file, pinned by `bash_history_survives_clear`), and both render one recording.
- `DriftPolicy.OFF` docstrings (`load`, `from_state`, `install_fingerprints`) no longer claim to "evict the snapshot cache": OFF drops the restored RAM cache entries, and a Redis cache is never restored from a snapshot, so there is nothing to drop and its `evict_paths` no-op is correct.
- Item 4 of #1024 (per-resource index config being discarded) is fixed by #1042.

Refs #1017, #1019, #1024.

## Integration coverage

The lifecycle battery (`integ/lifecycle/`, python + node + browser hosts) gains two host-side ops, `snapshot` and `checkout` (a `to_state_dict` / `apply_state_dict` round trip onto the running workspace), and its coded `RulePolicy` gains a `vars` rule that refuses `pre_session` for the names it lists. The new case `checkout_cannot_smuggle_a_variable_past_the_session_gate` exports `SECRET`, snapshots, unsets it, registers the rule, and expects the checkout to fail `EACCES` with the variable still unset; unregistering the rule lets the same checkout through and the variable comes back. All three hosts pass it, and removing the `gate_restored_vars` call makes step 5 fail, so the case does falsify the bug.

Not covered by integ, on purpose: the unmatched-mount warning is a log line and the redacted-alias override check needs a MinIO alias config in a snapshot, both stay unit-tested.

## After the Codex review

- **Every session table and the env template are vetted before anything lands.** `_gate_restored_state` / `gateRestoredState` run first in `apply_state_dict` / `applyStateDict`: each table is parsed and gated, then the template (under the id the restore makes the default), ahead of every mount's `load_state`, the default-session adoption and any session write. A snapshot with several sessions used to be gated one table at a time as each landed, so a refusal on a later session left the earlier ones overwritten, the default identity adopted and the mounts loaded, and a close then persisted that half-state. Two new tests per language (a refused later session, a refused template) assert the target is exactly as it was; both fail on the previous commit.
- **TypeScript reports a live-only snapshot mount with no mount at its prefix.** The exact-prefix lookup now runs before the `resourceStateRequiresOverride` skip, as Python's does, so a renamed remote or config-backed mount, the case the report exists for, is no longer silent. The `loadState` skip itself stays for a mount that is present. Pinned by a vitest case with `needs_override` mounts at a renamed and a matching prefix and a `loadState` spy on the matching one.

### CODE REVIEW

**Comment by zechengz** [comment]
@codex review

File: python/mirage/workspace/snapshot/state.py [hunk 1]
```diff
@@ -365,6 +382,7 @@ async def _restore_sessions(ws, state: dict[str, Any]) -> None:
                 # the replace_from_snapshot contract below.
                 session = ws._session_mgr.get(sid)
         fields = Session.from_dict(s_data)
+        await gate_restored_vars(ws.policies, sid, fields.vars)
```
[the remark this answers is not in the document]
Comment (bytecii) [reply python/mirage/workspace/snapshot/state.py]: Fixed in ddb07caf5. `apply_state_dict` / `applyStateDict` now start with `_gate_restored_state` / `gateRestoredState`: every session table is parsed and gated, then the env template (under the id the restore makes the default), before any mount's `load_state`, the default-session adoption or a session write. A refusal leaves the workspace exactly as it was, so a close persists nothing half-restored. Pinned by `test_a_refused_session_table_leaves_the_workspace_untouched` and `test_a_refused_env_template_lands_no_session` plus their vitest twins; all four fail on the previous commit.

File: typescript/packages/core/src/workspace/snapshot/state.ts [hunk 2]
```diff
@@ -403,7 +404,14 @@ export async function applyStateDict(ws: Workspace, state: WorkspaceStateDict):
     // workspace does not mount is skipped, never resolved to an
     // ancestor mount (which would load state into the wrong resource).
     const mount = ws.registry.tryMountForPrefix(m.prefix)
-    if (mount === null) continue
+    if (mount === null) {
+      // Said out loud: a renamed or missing mount otherwise left no trace.
+      console.warn(
+        `Workspace.load: snapshot mount ${m.prefix} has no mount at that prefix in this ` +
+          `workspace; its state was not restored`,
+      )
```
[the remark this answers is not in the document]
Comment (bytecii) [reply typescript/packages/core/src/workspace/snapshot/state.ts:438]: Fixed in ddb07caf5. The exact-prefix lookup and its warning now run before the `resourceStateRequiresOverride` skip, as Python's loop does, so a live-only mount (`needs_override`, a redacted credential) at a prefix this workspace lacks is reported; the `loadState` skip itself stays for a mount that is present. Pinned by 'a live-only snapshot mount with no matching prefix is reported too', which marks both saved mounts `needs_override`, renames one, and spies `loadState` on the one that matched; it fails on the previous commit.

**Comment by zechengz** [comment]
@codex review

File: python/mirage/workspace/snapshot/state.py [hunk 3]
```diff
@@ -344,7 +365,45 @@ async def _restore_nodes(ws, state: dict[str, Any]) -> None:
...
+    Returns:
+        The parsed session tables, and the env template or None when
+        the snapshot carries none.
+    """
+    sessions = [
+        Session.from_dict(s_data)
+        for s_data in state.get(StateKey.SESSIONS, [])
+    ]
+    for fields in sessions:
+        await gate_restored_vars(ws.policies, fields.session_id, fields.vars)
```
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/workspace/snapshot/state.py:418]: Fixed in 5e33c6109. The snapshot's own profile never lands: `_restore_sessions` / `restoreSessions` copy cwd, vars and mount modes onto the live session and nothing else, so for an id the target already has, the gate judged the table under the live session's script, which is the one it keeps. The gap was a session the restore creates: `script_of` answers for an unknown id with the default profile's script, so that is what the table was judged under, while the bare manager `create()` then landed the session under no profile at all. The created session is now narrowed under the document's default profile before its table lands, as `create_session` with no name does, so the policy the gate judged is the policy the session runs under, in both languages; an existing session keeps its own as before. Pinned by `test_a_session_the_restore_creates_runs_under_the_default_profile` and its vitest twin, which fail on the previous commit.

File: python/mirage/workspace/snapshot/state.py [hunk 4]
```diff
@@ -310,23 +317,37 @@ async def apply_state_dict(ws, state: dict[str, Any]) -> None:
     Workspace must already have its mounts constructed via the args
     from build_mount_args. This function is purely additive — it does
     not construct anything.
+
+    Every session table and the env template clear the target's
+    ``pre_session`` gate first (``_gate_restored_state``), before any
```
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/workspace/snapshot/state.py:326]: Fixed in 5e33c6109. The clear moved behind the gate: `apply_state_dict(ws, state, replace_cache=True)` / `applyStateDict(ws, state, { replaceCache: true })` drops the live cache after `_gate_restored_state` has passed and ahead of the mounts' `load_state`, and `checkout` and `restore` in both languages dropped their own clears, so a refused checkout leaves the cache with the rest of the workspace. A fresh `Workspace.load` keeps not clearing, since its TypeScript options can name an external cache the snapshot never touches. Pinned by `test_a_refused_checkout_leaves_the_live_cache_alone`, `test_a_refused_restore_leaves_the_live_cache_alone` and their vitest twins in `api.test.ts` and `stateDiff.test.ts`; all fail on the previous commit.

File: python/mirage/workspace/snapshot/state.py [hunk 5]
```diff
@@ -344,7 +383,53 @@ async def _restore_nodes(ws, state: dict[str, Any]) -> None:
...
+        await gate_restored_vars(ws.policies, fields.session_id, fields.vars)
+    seed = state.get(StateKey.ENV)
+    if not seed:
+        return sessions, None
+    default_sid = state.get(StateKey.DEFAULT_SESSION_ID)
+    seed_vars = vars_from_fields(seed)
+    await gate_restored_vars(
+        ws.policies,
+        ws._session_mgr.default_id if default_sid is None else default_sid,
+        seed_vars)
```
[the remark this answers is not in the document]
Comment (zechengz) [reply python/mirage/workspace/snapshot/state.py:427]: The env template is workspace-level state, not per-session state, and it is not gated per session anywhere, by design.

`ws.env = {...}` writes the default session's vars directly (the `SessionManager.env` setter) and fires no `pre_session` gate at all. Construction seeds the template onto the default session ungated, and `SessionManager.create` copies it into every created session ungated; a restrictive profile merges its own env over it (`apply_profile`) but never clears what was inherited. Gating the restored template once, under the id a template write lands on, is therefore already stricter than either of those paths, and it is the only place a foreign snapshot's template is vetted at all. The restored session *tables* are each gated under their own session id in the loop just above this call, so no restored session receives an unvetted variable.

The suggested change gates the seed vars as they are copied into each session, which is inside `create()`, the same door the deployment's own template flows through. It would make `create_session("x", profile=P)` raise `PolicyDenied` whenever P denies a name the deployment legitimately set in its env block, so a restored template would be stricter than a constructed one and a session could no longer inherit the workspace env it is documented to start from. Inheriting a starting environment is deliberately ungated for every session; only the writes a shell line or a restore performs are gated, which is what `gate_restored_vars` does here.

Not changing this one.
